presto-native-execution/presto_cpp/main/http/filters/EndpointLatencyFilter.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "presto_cpp/main/http/filters/EndpointLatencyFilter.h"
+
+namespace facebook::presto::http::filters {
+
+EndpointLatencyFilter::EndpointLatencyFilter(
+    proxygen::RequestHandler* upstream,
+    std::shared_ptr<std::unordered_map<
+        proxygen::HTTPMethod,
+        std::vector<std::unique_ptr<EndPoint>>>> endpoints)
+    : Filter(upstream), endpoints_(endpoints) {}
+
+// static
+void EndpointLatencyFilter::updateLatency(
+    const std::string& endpoint,
+    uint64_t latencyUs) {
+  metricMap().withWLock(
+      [&](std::unordered_map<std::string, EndPointMetrics>& map) {
+        auto itr = map.find(endpoint);
+        if (itr != map.end()) {
+          auto& metrics = itr->second;
+          metrics.maxLatencyUs = std::max(metrics.maxLatencyUs, latencyUs);
+          ++metrics.count;
+        } else {
+          map.emplace(endpoint, EndPointMetrics{endpoint, latencyUs, 1});
+        }
+      });
+}
+
+// static
+std::vector<EndpointLatencyFilter::EndPointMetrics>
+EndpointLatencyFilter::retrieveLatencies() {
+  std::vector<EndpointLatencyFilter::EndPointMetrics> result;
+  metricMap().withWLock([&](std::unordered_map<
+                            std::string,
+                            EndpointLatencyFilter::EndPointMetrics>& map) {
+    result.reserve(map.size());
+    for (const auto& pair : map) {
+      result.push_back(pair.second);
+    }
+    map.clear();
+  });
+  std::sort(result.begin(), result.end(), [](const auto& a, const auto& b) {
+    return a.maxLatencyUs > b.maxLatencyUs;
+  });
+  return result;
+}
+
+void EndpointLatencyFilter::onRequest(
```

Comment:
Is this method thread-safe? Is endpoints_ const? This object is created per each http msg?

Reply:
Yeah it is one EndpointLatencyFilter per request so thread-safe. 'endpoints_' should not be modified so yeah we can make it const.